pallets/creditcoin/src/tests.rs

```diff
@@ -341,19 +376,60 @@ fn register_address_should_work() {
 }
 
 #[test]
-fn register_address_pre_existing() {
+fn register_address_should_fail_to_reregister_external_address_to_same_account() {
 	ExtBuilder::default().build_and_execute(|| {
-		let (who, address, ownership_proof, _) = generate_address_with_proof("owner");
+		let (who, _) = generate_random_account("owner");
+		let (external_address, external_keypair) =
+			generate_random_external_address("external owner");
+		let ownership_proof = build_proof_of_ownership(who.clone(), external_keypair);
 		let blockchain = Blockchain::Rinkeby;
+		// Register external address to account
 		assert_ok!(Creditcoin::register_address(
 			Origin::signed(who.clone()),
 			blockchain.clone(),
-			address.clone(),
+			external_address.clone(),
 			ownership_proof.clone()
 		));
+		// Try registering again to same account and fail
+		assert_noop!(
+			Creditcoin::register_address(
+				Origin::signed(who),
+				blockchain,
+				external_address,
+				ownership_proof
+			),
+			crate::Error::<Test>::AddressAlreadyRegisteredByCaller
+		);
+	})
+}
 
+#[test]
+fn register_address_should_fail_to_reregister_external_address_to_new_account() {
+	ExtBuilder::default().build_and_execute(|| {
+		let (who_first, _) = generate_random_account("first account");
+		let (who_second, _) = generate_random_account("second account");
+		let (external_address, external_keypair) =
+			generate_random_external_address("external account");
+		let blockchain = Blockchain::Rinkeby;
+
+		// Register external address to first account
+		let first_ownership_proof =
+			build_proof_of_ownership(who_first.clone(), external_keypair.clone());
+		assert_ok!(Creditcoin::register_address(
+			Origin::signed(who_first),
+			blockchain.clone(),
+			external_address.clone(),
+			first_ownership_proof.clone()
+		));
+		// Try registering to a second account and fail
+		let second_ownership_proof = build_proof_of_ownership(who_second.clone(), external_keypair);
```

Comment:
Similar question here.

It looks like `generate_address_with_proof()` can still be called twice and in addition only need `build_proof_of_ownership()` to update `second_ownership_proof`.

Reply:
I guess you could... but it would confusing to me as to what's happening because you'd be throwing away a generated proof.
I think it's much clearer to have small functions that only give you what you need.

More importantly, I find it helpful to describe the external address and internal accounts as separate entities as they may (and probably will/should) come from different key pairs.

I was actually very confused until I realized the helper function assumes both the external address and internal account come from the same keypair.